CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2010-2023 Australian Signals Directorate
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package au.gov.asd.tac.constellation.utilities.text;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import javafx.scene.control.Label;
+import javafx.scene.control.ListView;
+import javafx.scene.input.MouseEvent;
+import javafx.stage.Popup;
+import org.apache.commons.lang3.StringUtils;
+import org.languagetool.JLanguageTool;
+import org.languagetool.MultiThreadedJLanguageTool;
+import org.languagetool.language.AustralianEnglish;
+import org.languagetool.rules.Rule;
+import org.languagetool.rules.RuleMatch;
+/**
+ * Handles the SpellChecking functions of SpellCheckingTextArea. SpellChecker
+ * evaluates incorrect words/phrases and pops up the suggestions when the user
+ * prompts
+ *
+ * @author Auriga2
+ */
+public final class SpellChecker {
+
+    private final SpellCheckingTextArea textArea;
+    private static final List<String> misspells = new ArrayList<>();
+    private List<RuleMatch> matches = new ArrayList<>();
+    private int indexOfMisspelledTextUnderCursor;       // position of the current misspelled text in misspells list
+    private final ListView<String> suggestions = new ListView<>(FXCollections.observableArrayList());
+    private final JLanguageTool langTool;
+    private Popup popup;
+    private boolean turnOffSpellChecking = false;
+    private int startOfMisspelledTextUnderCursor;
+    private int endOfMisspelledTextUnderCursor;
+    private static final Logger LOGGER = Logger.getLogger(SpellChecker.class.getName());
+
+    public SpellChecker(final SpellCheckingTextArea spellCheckingTextArea) {
+        textArea = spellCheckingTextArea;
+        langTool = new MultiThreadedJLanguageTool(new AustralianEnglish()); //Can pass in the language when supporting multiple languages
+
+        initialize();
+    }
+
+    /**
+     * set a listener on the list views selection model to get the text of the
+     * selected row.
+     */
+    private void initialize() {
+        for (Rule rule : langTool.getAllRules()) {
+            if (rule.getId().equals("UPPERCASE_SENTENCE_START")) {
+                langTool.disableRule(rule.getId());
```

Comment:
assuming the rule id is unique, could be worth adding a break after this since presumably every other rule will return false at the if condition

Reply:
Using the same loop to keep track of spellingCheckRule now. Hence cannot break now. But wondering whether to use 2 loops with break instead.?